### PR TITLE
Capability to instantiate an object based on the String class name. I…

### PR DESCRIPTION
**Pull request by fpirson**
…t is usefull when you want to set the MetricsTackerFactory from a property

### CODE REVIEW

**Comment by roidelapluie** [comment]
@brettwooldridge we are really interested in getting this merged. Can you please have a look?

File: src/main/java/com/zaxxer/hikari/util/PropertyElf.java [hunk 1]
```diff
@@ -146,7 +146,14 @@ else if (paramClass == String.class) {
             writeMethod.invoke(target, propValue.toString());
          }
          else {
-            writeMethod.invoke(target, propValue);
+            try {
+               LOGGER.info("Try to create a new instance of \"{}\"", propValue.toString());
```
Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/util/PropertyElf.java]: Log at debug level.

File: src/main/java/com/zaxxer/hikari/util/PropertyElf.java [hunk 2]
```diff
@@ -146,7 +146,14 @@ else if (paramClass == String.class) {
...
+               writeMethod.invoke(target, Class.forName(propValue.toString()).newInstance());
+            }catch(InstantiationException | ClassNotFoundException e){
```
Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/util/PropertyElf.java]: Formatting.  Space after ``}``, before ``(``, after ``)``.

File: src/main/java/com/zaxxer/hikari/util/PropertyElf.java [hunk 3]
```diff
@@ -146,7 +146,14 @@ else if (paramClass == String.class) {
...
+               LOGGER.info("Class \"{}\" not found or could not instantiate it (Default constructor)", propValue.toString());
```
Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/util/PropertyElf.java]: Log at debug level.

File: src/test/java/com/zaxxer/hikari/mocks/TestObject.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package com.zaxxer.hikari.mocks;
+
+public class TestObject{
```
Comment (brettwooldridge) [inline src/test/java/com/zaxxer/hikari/mocks/TestObject.java]: Formatting.  Space before ``{``.  Throughout the file.

File: src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import com.zaxxer.hikari.mocks.TestObject;
+import java.util.Properties;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotSame;
+import static org.junit.Assert.fail;
+
+
```
Comment (brettwooldridge) [inline src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java]: Remove this blank line.

File: src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+public class PropertyElfTest {
+
+   @Test
+   public void setTargetFromProperties() throws Exception{
```
Comment (brettwooldridge) [inline src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java]: Space before ``{``.

File: src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+   @Test
+   public void setTargetFromPropertiesNotAClass() throws Exception {
+      Properties properties = new Properties();
+      properties.setProperty("string", "aString");
+      properties.setProperty("testObject", "it is not a class");
+      TestObject testObject = new TestObject();
+      try {
+         PropertyElf.setTargetFromProperties(testObject, properties);
+         fail("Could never come here");
+      }catch (RuntimeException e){
```
Comment (brettwooldridge) [inline src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java]: Formatting.

File: src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+         assertEquals("argument type mismatch", e.getCause().getMessage());
+      }
+   }
+
+
+
+
```
Comment (brettwooldridge) [inline src/test/java/com/zaxxer/hikari/util/PropertyElfTest.java]: Remove these blank lines.

**Comment by fpirson** [comment]
Corrections done

**Pull request merged**